<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.7.32">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Estimating the binding affinity of the interaction between tandem beta-zippers and fibronectin – RJBingham</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js" type="module"></script>
<script src="site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-37eea08aefeeee20ff55810ff984fec1.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap-7baaf987bdc31585216d969921a59bec.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="styles.css">
</head>

<body class="nav-fixed quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">RJBingham</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="./index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./research.html"> 
<span class="menu-text">Research</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./teaching.html"> 
<span class="menu-text">Teaching</span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#extracting-the-amino-acid-sequences" id="toc-extracting-the-amino-acid-sequences" class="nav-link active" data-scroll-target="#extracting-the-amino-acid-sequences">Extracting the amino acid sequences</a>
  <ul class="collapse">
  <li><a href="#staphylococcus-aureus-fibronectin-binding-protein-a-repeat-1-fnbpa-1" id="toc-staphylococcus-aureus-fibronectin-binding-protein-a-repeat-1-fnbpa-1" class="nav-link" data-scroll-target="#staphylococcus-aureus-fibronectin-binding-protein-a-repeat-1-fnbpa-1"><em>Staphylococcus aureus</em> Fibronectin-binding Protein A Repeat 1 (FnBPA-1)</a></li>
  <li><a href="#streptococcal-fibronectin-binding-protein-i-sfbi" id="toc-streptococcal-fibronectin-binding-protein-i-sfbi" class="nav-link" data-scroll-target="#streptococcal-fibronectin-binding-protein-i-sfbi">Streptococcal fibronectin-binding protein I (SfbI)</a></li>
  <li><a href="#fibronectin-domains-2f1-to-5f1" id="toc-fibronectin-domains-2f1-to-5f1" class="nav-link" data-scroll-target="#fibronectin-domains-2f1-to-5f1">Fibronectin domains 2F1 to 5F1</a></li>
  </ul></li>
  <li><a href="#generating-a-model-of-the-tandem-beta-zipper-using-alphafold3" id="toc-generating-a-model-of-the-tandem-beta-zipper-using-alphafold3" class="nav-link" data-scroll-target="#generating-a-model-of-the-tandem-beta-zipper-using-alphafold3">Generating a model of the tandem beta-zipper using AlphaFold3</a></li>
  <li><a href="#calculating-binding-affinity-using-prodigy" id="toc-calculating-binding-affinity-using-prodigy" class="nav-link" data-scroll-target="#calculating-binding-affinity-using-prodigy">Calculating binding affinity using PRODIGY</a></li>
  <li><a href="#correlation-between-prediction-and-experiment" id="toc-correlation-between-prediction-and-experiment" class="nav-link" data-scroll-target="#correlation-between-prediction-and-experiment">Correlation between prediction and experiment</a></li>
  <li><a href="#umbrella-sampling" id="toc-umbrella-sampling" class="nav-link" data-scroll-target="#umbrella-sampling">Umbrella Sampling</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Estimating the binding affinity of the interaction between tandem beta-zippers and fibronectin</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<style>
blockquote code,
blockquote pre {
  background-color: #d8d8d8;
  border: 1px solid #ccc;
  border-radius: 6px;
  padding: 0.75em;
  box-shadow: 0 2px 4px rgba(0,0,0,0.1);
}
</style>
<p>This guide is written for project students at undergraduate or MSc level.</p>
<p>The aim is to extract the amino acid sequences for the target proteins, build an Alphafold model of the complex, then submit this to PRODIGY to estimate the binding affinity. We start with an introduction to coordinate files. ## Introduction to coordinate files</p>
<p>Several experimental structures (from NMR and X-ray crystallography) are available of bacterial peptides in complex with fibronectin. The structure files can be downloaded from the Protein Data Bank (PDB). For example, see the “Crystal structure of the second and third fibronectin F1 modules in complex with a fragment of <em>Streptococcus pyogenes</em> SfbI-5”, PDB accession code <a href="https://www.rcsb.org/structure/3ZRZ">3ZRZ</a>. Click on Download Files, then choose either the PDBx/mmCIF Format, or the Legacy PDB Format.</p>
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><img src="images/PDB_screenshot.png" class="img-fluid figure-img" style="width:30.0%"></p>
<figcaption>screenshot of the PDB)</figcaption>
</figure>
</div>
<p>This is a simple text file. Open it in a text editor (e.g.&nbsp;Wordpad) and scroll down to see the 3D coordinates of each atom. You will see this to specify the details for atom 1 of 3ZRZ.pdb</p>
<pre><code>ATOM      1  N   GLU A  63      -9.834 -21.054   6.812  1.00 33.20           N </code></pre>
<p>Reading across, the columns are:</p>
<ul>
<li>atom number (1),</li>
<li>The element (N = Nitrogen)</li>
<li>Three letter code for the type of residue (Glu = Glutamic acid)</li>
<li>Single letter to specify the chain ID (A = Chain A)</li>
<li>The residue number within that chain (Residue 63)</li>
<li>Cartesian coordinates x, y, z (-9.834 -21.054 6.812)</li>
<li>The occupancy - a number between zero to 1 (1.00)<br>
</li>
<li>The <a href="https://pubmed.ncbi.nlm.nih.gov/30698416/">B-factor</a> of this atom (33.20)</li>
</ul>
<p>There are several programs available to visualise coordinate files in this format (<a href="https://www.pymol.org/">PyMOL</a>, <a href="https://www.ks.uiuc.edu/Research/vmd/">VMD</a>, RASMOL), but these will not be covered here.</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Quick Tip
</div>
</div>
<div class="callout-body-container callout-body">
<p>If you open a coordinate file in PyMOL, your changes (e.g.&nbsp;colouring, deleting atoms etc.) are not automatically saved to the PDB file. The text file will remain unaltered. You can generate a new PDB file using PyMOL using File-Export Molecule, but you should always use a new filename.</p>
</div>
</div>
<p>There are numerous advantages of using coordinate files from experimental methods such as X-ray crystallography. The primary advantage is the accuracy of the structure - we can have high confidence in the precise spatial arrangement of all atoms in the coordinate file. However, there are some things we need to aware of:</p>
<ul>
<li>loop regions may be disordered. The amino acids in these flexible regions are often missing in the coordinate files, therefore there is a break, or gap in the polypeptide.</li>
<li>Some amino acids may have been modelled in two or more conformations (alternative conformations). Or, part of the side-chain might be missing.</li>
<li>The coordinate file (the asymmetric unit) may have more than one copy of the molecule of interest. For example, there are six copies of the biological relevant unit in <a href="https://www.rcsb.org/structure/2RL0">2RL0</a>. You can visualise these if you colour by chain in PyMOL. There are subtle differences between these six structures.<br>
</li>
<li>The coordinate files will often contain other non-protein components. If you look at <a href="https://www.rcsb.org/structure/3ZRZ">3ZRZ</a>, you will see a glycerol molecule, a sulphate ion and many water molecules. In addition, the bacterial peptide is acetylated at the N-terminus (ACE residue)</li>
</ul>
<p>To calculate an accurate binding affinity between two molecules, we need accurate coordinate files that include all the amino acids. These missing (or extra) parts will need to be corrected.</p>
<p>One option would be to edit the text file to remove these non-protein atoms, then rebuild and model the missing loops and side-chains. However, in 2021, Alphafold2 was released, which, for the first time made it possible to accurately predict the 3D-structure of a protein based only on the amino acid sequence <a href="https://www.nature.com/articles/s41586-021-03819-2">(Jumper et al., 2021)</a>. The accuracy of these protein models has been shown to be excellent, especially where there is a prior structure in the PDB with high sequence identity to the target. Then, in 2024 Alphafold3 was released <a href="https://www.nature.com/articles/s41586-024-07487-w">(Abramson et al., 2024)</a> allowing the accurate prediction of complexes, for example the interaction between two distinct proteins. The Alphafold server is available <a href="https://alphafoldserver.com/welcome">here</a>. You will need a google account to use it. You will notice from this server that the input is simply an amino acid sequence in one-letter code.</p>
<section id="extracting-the-amino-acid-sequences" class="level2">
<h2 class="anchored" data-anchor-id="extracting-the-amino-acid-sequences">Extracting the amino acid sequences</h2>
<p>In order to generate Alphafold models, we need to extract the correct amino acid sequences of the target proteins. In the example, we will find the amino acid sequences that correspond to the interaction between two different bacterial peptides (SfbI and FnBPA1) and the 2nd and 3rd F1 domains of human fibronectin (2F1-5F1).</p>
<section id="staphylococcus-aureus-fibronectin-binding-protein-a-repeat-1-fnbpa-1" class="level3">
<h3 class="anchored" data-anchor-id="staphylococcus-aureus-fibronectin-binding-protein-a-repeat-1-fnbpa-1"><em>Staphylococcus aureus</em> Fibronectin-binding Protein A Repeat 1 (FnBPA-1)</h3>
<p>FnBPA is a bacterial peptide from <em>Staphylococcus aureus</em> with 11 Fn-binding repeats. The interaction between FnBPA and fibronectin is descibed in <a href="https://doi.org/10.1074/jbc.M703063200">Meenan et al, 2007</a>.</p>
<p>Notice from Table 1 in <a href="https://www.jbc.org/article/S0021-9258(20)74663-6/fulltext">Meenan et al</a>, that FnBPA-1 corresponds to residues 512-550. Go to the uniprot entry for FnBPA <a href="https://www.uniprot.org/uniprotkb/P14738/entry">P14738</a>, find the amino acids 512-550, and paste them into a text file, as below.</p>
<pre><code>&gt;FnBPA1(512-550)
GPIIQNNKFEYKEDTIKETLTGQYDKNLVTTVEEEYDSS</code></pre>
<p>Notice that this is identical to the amino acid sequence for FnBPA1 in Figure 1, part B.</p>
</section>
<section id="streptococcal-fibronectin-binding-protein-i-sfbi" class="level3">
<h3 class="anchored" data-anchor-id="streptococcal-fibronectin-binding-protein-i-sfbi">Streptococcal fibronectin-binding protein I (SfbI)</h3>
<p>SfbI is bacterial peptide from <em>Streptococcus pyogenes</em>. The intereaction between SfBI and fibronectin is described in <a href="https://www.nature.com/articles/nature01589">Schwarz-Linek et al., (2003)</a>. The five repeats of SfBI are numbered 1-5 (top of Figure 3). In this example, we are interested in the interaction between 2F1-3F1 and repeat number 5. On Figure 3 you can see that the researchers have measured the binding affinity for 3 fragments of SfbI-5. The middle one with the sequence TGMSGFSETVTIVEDTRP binds to 2F1-3F1 with a measured binding affinity of 3.6 micromolar. The figure is an image, so we cannot cut and paste this peptide sequence from here. Typing it out is error-prone and time consuming. If you search the <a href="https://www.uniprot.org/">uniprotKB</a> for SfbI. You will see many entries from different isolates, many of which will have sequence variations from what was experimentally measured in <a href="https://www.nature.com/articles/nature01589">Schwarz-Linek et al., (2003)</a>. I could not find SfbI in the reference proteome on Uniprot, possibly because the protein is annotated with a different name. Instead, we will search the <a href="https://www.ncbi.nlm.nih.gov/protein/">NCBI</a>. Type “SfbI” in to the search box, then select “RefSeq” on the left side to only show sequences for reference proteomes:</p>
<p><img src="images/NCBI_search.png" class="img-fluid" style="width:60.0%"></p>
<p><img src="images/RefSeq.png" class="img-fluid" style="width:20.0%"></p>
<p>The top hit is <a href="https://www.ncbi.nlm.nih.gov/protein/WP_415244991.1">Fibronectin-binding protein PrtF1/SfbI</a> from <em>Streptococcus pyogenes</em>. Click on it and scroll to the amino acid sequence at the bottom. You now need to find the match between Figure 3 in <a href="https://www.nature.com/articles/nature01589">Schwarz-Linek et al., (2003)</a> and this sequence. You can use Control-F, but be aware of the spaces. Below you can see the region of SfbI-5 that binds to 2F1-3F1 (TGMSGFSETVTIVEDTRP). <img src="images/sequence_match.png" class="img-fluid"> Copy this sequence and paste it into a text file. Bookmark the Refseq above for easy access in the future.</p>
<pre><code>SfbI-5_middle
TGMSGFSETVTIVEDTRP</code></pre>
</section>
<section id="fibronectin-domains-2f1-to-5f1" class="level3">
<h3 class="anchored" data-anchor-id="fibronectin-domains-2f1-to-5f1">Fibronectin domains 2F1 to 5F1</h3>
<p>The Uniprot entry for human fibronectin is <a href="https://www.uniprot.org/uniprotkb/P02751/entry">P02751</a>. Click on Family and Domains, then scroll down to find the 2nd type-I domain (2F1), this is called “Fibronectin type-I 2” and it is residues 95-138. Scroll down further to the 5th type-I domain (5F1), this is called “Fibronectin type-I 5” and it is residues 229-273. Therefore, 2F1-5F1 must be all of these residues 95-273.</p>
<p>Scroll down to the <a href="https://www.uniprot.org/uniprotkb/P02751/entry#sequences">sequence section</a>. Select residues 95-273, copy and paste the sequence into a text editor and give it a sensible filename. Below, I have added a title in <a href="https://en.wikipedia.org/wiki/FASTA_format">FASTA format</a></p>
<pre><code>&gt;2F1-5F1
ETCFDKYTGNTYRVGDTYERPKDSMIWDCTCIGAGRGRISCTIANRCHEGGQSYKIGDTWRRPHETGGYMLECVCLGNGKGEWTCKPIAEKCFDHAAGTSYVVGETWEKPYQGWMMVDCTCLGEGSGRITCTSRNRCNDQDTRTSYRIGDTWSKKDNRGNLLQCICTGNGRGEWKCERH</code></pre>
</section>
</section>
<section id="generating-a-model-of-the-tandem-beta-zipper-using-alphafold3" class="level2">
<h2 class="anchored" data-anchor-id="generating-a-model-of-the-tandem-beta-zipper-using-alphafold3">Generating a model of the tandem beta-zipper using AlphaFold3</h2>
<p>Now you have the amino acid sequences in one-letter code, it is straightforward to generate a 3D model of the protein-protein interaction. Go to the <a href="https://alphafoldserver.com/welcome">Alphafold server</a>, paste in the sequence of fibronectin (2F1-5F1), click on “Add entity”, then paste in the corresponding sequence of the bacterial peptide (FnBPA1). Click on “Continue and preview job”. Once done, your completed jobs will appear as a list on the lower part of the screen. Click on the 3 dots on the right-side and download the results as a zip file. Extract the zip file, then look for the five models (the ones ending in .cif). You can visualise this in PyMOL. You should see something similar to the figure below. <img src="images/2F15F1-FnBPA1wide.png" class="img-fluid" alt="Alphafold3 model of FnBPA-1 interaction with 2F1-5F1"> Compare this model of FnBPA-1 interacting with 2F1-5F1 with the two X-ray crystal structures (separated by the dashed line) shown in Figure 1A in <a href="https://www.pnas.org/doi/full/10.1073/pnas.0803556105">Bingham et al., (2008)</a>. Notice that the fold of the individual Fn domains is the same, but the relative orientation of the domains is slightly different. The relative orientation shown in the 2008 paper was estimated based on the superposition with other Fn typeI-typeI domain interactions. The orientation in the Alphafold model is likely to be more accurate (have higher occupancy). It is worth noting that extended multidomain systems such as this have a degree of flexibility, so the model we generate is just a snapshot of one possible orientation.</p>
</section>
<section id="calculating-binding-affinity-using-prodigy" class="level2">
<h2 class="anchored" data-anchor-id="calculating-binding-affinity-using-prodigy">Calculating binding affinity using PRODIGY</h2>
<p>The <a href="https://rascar.science.uu.nl/prodigy/">PRODIGY</a> webserver accepts files in .pdb format. However, the output from the Alphafold server is a .cif coordinate file which has a slightly different format. One way to convert .cif to .pdb is to use PyMOL.</p>
<ul>
<li>As above, download the results from Alphafold and extract the zip file. You should see 5 models with the .cif extension</li>
<li>Open one of the files in PyMOL.</li>
<li>Click on File -&gt; Export Molecule</li>
<li>On the Selection, use the dropdown menu to choose the correct model</li>
<li>Click Save</li>
<li>At the bottom-right, there is a menu that allows you to change the format of the output. Choose ‘PDB’. (not PDBx/mmcif)</li>
<li>Type in a suitable name and click Save</li>
</ul>
<p>Now, submitting this to <a href="https://rascar.science.uu.nl/prodigy/">PRODIGY</a> is straightforward. Simply upload the pdb coordinate file and fill out the boxes with appropriate values then click ‘Submit PRODIGY’. Interactor 1 is fibronectin, (A) and Interactor 2 is the bacterial peptide (B).</p>
<p>Within minutes, you see a result similar to below.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/PRODIGY_result.png" class="img-fluid figure-img"></p>
<figcaption>PRODIGY output</figcaption>
</figure>
</div>
<p>The experimental Kd value for this interaction is 4.8 ±0.6 nM <a href="https://www.jbc.org/article/S0021-9258(20)74663-6/fulltext">(Meenan et al., 2007)</a>. However, the value from PRODIDGY is only 0.0000039 nM. (PRODIGY reports a value in moles/L so mulitiply the value above by 10^9 to convert 3.9e-15 M to 3.9e-6 nM) Why is there such a large discrepancy? The value from PRODIGY is over 1 million times smaller than the experimentally measured value. To answer this, you need to consider the dynamics of this particular type of protein-protein interaction and the limitations of the PRODIGY method (see the PRODIGY publications). The bacterial peptide transitions from being disordered in the unbound state to an extended beta-strand in the bound state. Compare this scenario to the training set of protein-protein complexes used to develop PRODIGY.</p>
</section>
<section id="correlation-between-prediction-and-experiment" class="level2">
<h2 class="anchored" data-anchor-id="correlation-between-prediction-and-experiment">Correlation between prediction and experiment</h2>
<p>If you repeat the above analysis for other tandem beta-zipper interactions, and you compare the predicted Kd values to those from ITC experiments, you may find there is a linear correlation similar to below. Questions: - What is the gradient of this linear model, is that informative? - Does the linear model have any predictive value? - In previous analysis, BBK32 appears as an outlier. Why? - Can this method be extended to other protein-protein interactions that involve disorder to order transitions and/or beta-strand addition</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/FnBPs_correlation.png" class="img-fluid figure-img"></p>
<figcaption>FnBPs_correlation.png</figcaption>
</figure>
</div>
</section>
<section id="umbrella-sampling" class="level2">
<h2 class="anchored" data-anchor-id="umbrella-sampling">Umbrella Sampling</h2>
<p>The method presented above clearly underestimates the binding affinity of the tandem beta-zipper. A more accurate computational method to calculate the thermodynamic parameters is <a href="./umbrella_sampling_tutorial.html">umbrella sampling</a>.</p>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
        const codeEl = trigger.previousElementSibling.cloneNode(true);
        for (const childEl of codeEl.children) {
          if (isCodeAnnotation(childEl)) {
            childEl.remove();
          }
        }
        return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp('/' + window.location.host + '/');
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->




</body></html>