<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.7.32">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Umbrella sampling SfbI5 bound to 2F1-3F1 – RJBingham</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js" type="module"></script>
<script src="site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-37eea08aefeeee20ff55810ff984fec1.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap-7baaf987bdc31585216d969921a59bec.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

<link rel="stylesheet" href="styles.css">
</head>

<body class="nav-fixed quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">RJBingham</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="./index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./research.html"> 
<span class="menu-text">Research</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./teaching.html"> 
<span class="menu-text">Teaching</span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#download-the-input-files-and-charmm36-forcefield" id="toc-download-the-input-files-and-charmm36-forcefield" class="nav-link active" data-scroll-target="#download-the-input-files-and-charmm36-forcefield">Download the input files and Charmm36 forcefield</a></li>
  <li><a href="#generate-an-alphafold-model-of-sfbi-5-in-complex-with-2f1-3f1" id="toc-generate-an-alphafold-model-of-sfbi-5-in-complex-with-2f1-3f1" class="nav-link" data-scroll-target="#generate-an-alphafold-model-of-sfbi-5-in-complex-with-2f1-3f1">Generate an Alphafold model of SfbI-5 in complex with 2F1-3F1</a></li>
  <li><a href="#build-the-system" id="toc-build-the-system" class="nav-link" data-scroll-target="#build-the-system">Build the System</a></li>
  <li><a href="#solvate-and-neutralise" id="toc-solvate-and-neutralise" class="nav-link" data-scroll-target="#solvate-and-neutralise">Solvate and neutralise</a></li>
  <li><a href="#minimise-and-equilibrate" id="toc-minimise-and-equilibrate" class="nav-link" data-scroll-target="#minimise-and-equilibrate">Minimise and Equilibrate</a></li>
  <li><a href="#check-equilibration-of-temperature-and-pressure" id="toc-check-equilibration-of-temperature-and-pressure" class="nav-link" data-scroll-target="#check-equilibration-of-temperature-and-pressure">Check equilibration of temperature and pressure</a></li>
  <li><a href="#ns-of-unrestrained-equilibration" id="toc-ns-of-unrestrained-equilibration" class="nav-link" data-scroll-target="#ns-of-unrestrained-equilibration">50 ns of unrestrained equilibration</a></li>
  <li><a href="#check-equilibration-of-temperature-and-pressure-during-the-50-ns-simulation" id="toc-check-equilibration-of-temperature-and-pressure-during-the-50-ns-simulation" class="nav-link" data-scroll-target="#check-equilibration-of-temperature-and-pressure-during-the-50-ns-simulation">Check equilibration of temperature and pressure during the 50 ns simulation</a></li>
  <li><a href="#transfer-the-protein-to-a-larger-box-optional" id="toc-transfer-the-protein-to-a-larger-box-optional" class="nav-link" data-scroll-target="#transfer-the-protein-to-a-larger-box-optional">Transfer the protein to a larger box (optional)</a></li>
  <li><a href="#pulling-the-tandem-beta-zipper-apart" id="toc-pulling-the-tandem-beta-zipper-apart" class="nav-link" data-scroll-target="#pulling-the-tandem-beta-zipper-apart">Pulling the tandem beta zipper apart</a>
  <ul class="collapse">
  <li><a href="#define-the-pull-groups" id="toc-define-the-pull-groups" class="nav-link" data-scroll-target="#define-the-pull-groups">Define the pull groups</a></li>
  <li><a href="#check-the-distances-in-the-npt-equilibration-this-is-a-sanity-check-to-ensure-we-picked-the-correct-atoms" id="toc-check-the-distances-in-the-npt-equilibration-this-is-a-sanity-check-to-ensure-we-picked-the-correct-atoms" class="nav-link" data-scroll-target="#check-the-distances-in-the-npt-equilibration-this-is-a-sanity-check-to-ensure-we-picked-the-correct-atoms">check the distances in the npt equilibration (this is a sanity check to ensure we picked the correct atoms)</a></li>
  <li><a href="#add-restraints-to-maintain-the-structure-of-fn" id="toc-add-restraints-to-maintain-the-structure-of-fn" class="nav-link" data-scroll-target="#add-restraints-to-maintain-the-structure-of-fn">Add restraints to maintain the structure of Fn</a></li>
  <li><a href="#edit-the-pull.mdp" id="toc-edit-the-pull.mdp" class="nav-link" data-scroll-target="#edit-the-pull.mdp">Edit the pull.mdp</a></li>
  <li><a href="#generate-the-configurations-pulling" id="toc-generate-the-configurations-pulling" class="nav-link" data-scroll-target="#generate-the-configurations-pulling">Generate the configurations (pulling)</a></li>
  </ul></li>
  <li><a href="#measure-the-distances-between-the-two-cα-atoms-over-time" id="toc-measure-the-distances-between-the-two-cα-atoms-over-time" class="nav-link" data-scroll-target="#measure-the-distances-between-the-two-cα-atoms-over-time">Measure the distances between the two Cα atoms over time</a></li>
  <li><a href="#get-the-individual-gro-files-for-umbrella-sampling" id="toc-get-the-individual-gro-files-for-umbrella-sampling" class="nav-link" data-scroll-target="#get-the-individual-gro-files-for-umbrella-sampling">Get the individual gro files for umbrella sampling</a></li>
  <li><a href="#visual-check-of-the-starting-positions-in-pymol" id="toc-visual-check-of-the-starting-positions-in-pymol" class="nav-link" data-scroll-target="#visual-check-of-the-starting-positions-in-pymol">visual check of the starting positions in pymol</a></li>
  <li><a href="#run-the-umbrella-sampling" id="toc-run-the-umbrella-sampling" class="nav-link" data-scroll-target="#run-the-umbrella-sampling">Run the umbrella sampling</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Umbrella sampling SfbI5 bound to 2F1-3F1</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<p>This tutorial is incomplete, as we work through each stage I will edit this page to reflect what we actually do.</p>
<p>The aim of this guide is to use umbrella sampling to calculate the dissociation constant (Kd) for the interaction between 2F1-3F1 and a fragment of <a href="https://www.ncbi.nlm.nih.gov/protein/WP_415244991.1">SfbI</a> (SfbI-5 residues 574-591). The methods are based on Exercise 3 in the <a href="https://pmc.ncbi.nlm.nih.gov/articles/PMC11457149/">tutorial guide by Justin Lemkul</a>. We will start by downloading these files from github.</p>
<section id="download-the-input-files-and-charmm36-forcefield" class="level2">
<h2 class="anchored" data-anchor-id="download-the-input-files-and-charmm36-forcefield">Download the input files and Charmm36 forcefield</h2>
<p>Create a new folder, then run the commands below to download two folders “gmx_tutorials_jpcb” and “toppar_c36_releases”.</p>
<pre><code>git clone https://github.com/Lemkul-Lab/gmx_tutorials_jpcb.git
git clone https://github.com/Lemkul-Lab/toppar_c36_releases.git</code></pre>
<p>Go into the gmx_tutorials_jpcb/inputs, and run these commands below. The first command makes a copy of the ‘03_chignolin’ folder and renames it to ‘04_sfbi’. This will be the working directory. The second command goes up two levels, then copies the charmm36 forcefield to the working directory “04_sfbi”. This forcefield will then appear at number 1 on the list when you run pdb2gmx command.</p>
<pre><code>cp -r 03_chignolin 04_sfbi
cp -r ../../toppar_c36_releases/gromacs/charmm36-jul2024.ff/ 04_sfbi/</code></pre>
<p>You should now have these files and directories listed below. Check by running “ls -ltr”</p>
<pre><code>1uao_model1.pdb
find_windows.py
inputs
run_umbrella_sampling.sh
write_wham_files.py
charmm36-jul2024.ff</code></pre>
</section>
<section id="generate-an-alphafold-model-of-sfbi-5-in-complex-with-2f1-3f1" class="level2">
<h2 class="anchored" data-anchor-id="generate-an-alphafold-model-of-sfbi-5-in-complex-with-2f1-3f1">Generate an Alphafold model of SfbI-5 in complex with 2F1-3F1</h2>
<p>The fragment of <a href="https://www.ncbi.nlm.nih.gov/protein/WP_415244991.1">SfbI</a> that we will target is shown on Figure 3 in <a href="https://www.nature.com/articles/nature01589">Schwarz-Linek et al., 2003</a>. The amino acid sequence is shown below alongside the sequence of 2F1-3F1 from <a href="https://www.uniprot.org/uniprotkb/P02751/entry#family_and_domains">P02751</a>.</p>
<pre><code>&gt;SfbI5(2F1-3F1 binding region 574-591)
TGMSGFSETVTIVEDTRP
&gt;2F1-3F1 (P02751 95-182)
ETCFDKYTGNTYRVGDTYERPKDSMIWDCTCIGAGRGRISCTIANRCHEGGQSYKIGDTWRRPHETGGYMLECVCLGNGKGEWTCKPI</code></pre>
<p>Generate an Alphafold3 model of this complex as described <a href="./tandem_beta_prodigy.html">previously</a>. Save the structure in PDB format, then copy this file into the 04_sfbi folder.</p>
</section>
<section id="build-the-system" class="level2">
<h2 class="anchored" data-anchor-id="build-the-system">Build the System</h2>
<p>In the 04_sfbi folder, run pdb2gmx using the command below. This will add hydrogens and generate coordinate and topology files in GROMACS format.</p>
<pre><code>gmx pdb2gmx -f fold_sfbi5_2f13f1_model_0.pdb -ignh </code></pre>
<p>The program will look for forcefields in the current working directory that end with ‘.ff’. It should find the charmm36 forcefield from above, and this should be number 1 on the list.</p>
<p>Select 1 for charmm, the select 1 for the TIP3P water model.</p>
<p>Read the text output, paying careful attention to the start/end terminii and the linking of disulfide bonds. The output is “conf.gro”. You can check this in PyMOL. We now need to increase the simulation box size. -f conf.gro is the input file -o box.gro is the output file -bt cubic specifies that the box type will be cubic -box 10 specifies that the length of each side will be 10 nm -c centres the molecule in the box</p>
<pre><code>gmx editconf -f conf.gro -o box.gro -bt cubic -box 10 -c</code></pre>
<p>The output shows:</p>
<pre><code>Volume: 40.0493 nm^3, corresponds to roughly 18000 electrons
No velocities found
    system size :  5.728  2.729  2.563 (nm)
    diameter    :  5.749               (nm)
    center      : -0.004  0.002  0.059 (nm)
    box vectors :  5.728  2.729  2.562 (nm)
    box angles  :  90.00  90.00  90.00 (degrees)
    box volume  :  40.05               (nm^3)
    shift       :  5.004  4.998  4.941 (nm)
new center      :  5.000  5.000  5.000 (nm)
new box vectors : 10.000 10.000 10.000 (nm)
new box angles  :  90.00  90.00  90.00 (degrees)
new box volume  :1000.00               (nm^3)</code></pre>
</section>
<section id="solvate-and-neutralise" class="level2">
<h2 class="anchored" data-anchor-id="solvate-and-neutralise">Solvate and neutralise</h2>
<p>The ITC experiements for SfbI peptides binding to 2f1-3F1 were conducted at low ionic strength buffer (10 mM, <a href="https://www.nature.com/articles/nature01589">see reference 17</a>) pH7.4, 25 °C. We will use 0.15 M NaCl to be consistent with other experimental work and other simulations. The files in this tutorial (downloaded from github earlier) are already at 298 K (25 °C), so no changes are required here. The protonation is already consistent with neutral pH. We can specify a NaCl concentration of 0.15 M with ‘-conc 0.15’ as below:</p>
<pre><code>gmx solvate -cp box.gro -cs spc216.gro -p topol.top -o solv.gro
gmx grompp -f inputs/ions.mdp -c solv.gro -p topol.top -o ions.tpr
gmx genion -s ions.tpr -o solv_ions.gro -p topol.top -pname NA -nname CL -neutral -conc 0.15</code></pre>
<p>Choose 13 to replace the waters with NaCl.</p>
</section>
<section id="minimise-and-equilibrate" class="level2">
<h2 class="anchored" data-anchor-id="minimise-and-equilibrate">Minimise and Equilibrate</h2>
<p><strong>Energy Minimisation</strong> Energy minimisation removes steric clashes, bad contacts, and high-energy configurations that arise during system preparation. In this step, the system is not simulated over time; instead, atomic positions are iteratively adjusted to find a nearby local minimum on the potential energy surface.</p>
<p><strong>NVT equilibration (constant Number of particles, Volume, Temperature)</strong> This phase equilibrates the temperature of the system while keeping the volume fixed. The thermostat is active, but the box size does not change. The goal is to stabilise the kinetic energy and remove any artefacts from energy minimisation.</p>
<p><strong>NPT equilibration (constant Number of particles, Pressure, Temperature)</strong> Once temperature is stable, NPT allows the pressure to equilibrate. A barostat is applied so the box can change size and density. This step brings the system to the correct physical density (1 bar) for production simulations.</p>
<p>The following will perform energy minimisation, then equilibration.</p>
<pre><code>#minimise (less than a minute)
gmx grompp -f inputs/minim.mdp -c solv_ions.gro -p topol.top -o em.tpr
gmx mdrun -deffnm em -v

#100 ps of nvt equilibration (1 minute)
gmx grompp -f inputs/nvt.mdp -c em.gro -r em.gro -p topol.top -o nvt.tpr
gmx mdrun -deffnm nvt -v

#500 ps of npt equilibration (~3 minutes)
gmx grompp -f inputs/npt.mdp -c nvt.gro -t nvt.cpt -r nvt.gro -p topol.top -o npt.tpr
gmx mdrun -deffnm npt -v</code></pre>
</section>
<section id="check-equilibration-of-temperature-and-pressure" class="level2">
<h2 class="anchored" data-anchor-id="check-equilibration-of-temperature-and-pressure">Check equilibration of temperature and pressure</h2>
<p>The commands below will extract the values for pressure and temperature during the equilibration process and produce a graph of each.</p>
<pre><code>gmx energy -f npt.edr -o npt_pressure.xvg 
xmgrace npt_pressure.xvg  #fluctuations should be less than 200 bar. 
gmx energy -f npt.edr -o npt_temperature.xvg
xmgrace npt_temperature.xvg </code></pre>
</section>
<section id="ns-of-unrestrained-equilibration" class="level2">
<h2 class="anchored" data-anchor-id="ns-of-unrestrained-equilibration">50 ns of unrestrained equilibration</h2>
<p>The structure is a model from Alphafold and so it is possible that some aspects of the geometry or structure are not accurate. There could be unusual hydrogen bond lengths, strained geometry, artefacts in covalent bond lengths/angles, or the packing of amino acid side chains might not be optimal. Furthermore, the structure needs to be equilibrated with the GROMACS force field (with our chosen temperature/pressure etc.). Therefore, we will run unrestrained MD simulation for 50 ns. Navigate to the ‘inputs’ folder and open md.mdp. Notice that this will run 25000000 steps of 0.002 ps, totalling 50 ns. Do not change anything, all is good. Start the 50 ns of unrestrained MD using the commands below.</p>
<p>The flags below are: -f is the mdp instruction file<br> -c is the input file<br> -t is Full precision trajectory input file (binary snapshot of the integrator state with positions, velocities, box, RNG/integrator state, elapsed time/step)<br> -p is the topology input file<br> -o is output trr file<br></p>
<pre><code>gmx grompp -f inputs/md.mdp -c npt.gro -t npt.cpt -p topol.top -o md_50ns.tpr
gmx mdrun -v -deffnm md_50ns</code></pre>
</section>
<section id="check-equilibration-of-temperature-and-pressure-during-the-50-ns-simulation" class="level2">
<h2 class="anchored" data-anchor-id="check-equilibration-of-temperature-and-pressure-during-the-50-ns-simulation">Check equilibration of temperature and pressure during the 50 ns simulation</h2>
<pre><code>gmx energy -f md_50ns.edr -o md_50ns_pressure.xvg
xmgrace md_50ns_pressure.xvg
gmx energy -f md_50ns.edr -o md_50ns_temperature.xvg
xmgrace md_50ns_temperature.xvg </code></pre>
<p>Fluctations should be mostly less than 200 bar</p>
<p>Now we have a protein structure that is equilibrated with the gromacs force field under our chosen conditions, we can proceed. The starting structure should be stable, so that as we run the individual umbrella simulations, there should not be any major conformational changes occurring. The changes in free energy we measure will be dominated by the separation of the bacterial peptide from the fibronectin. Check the peptide interaction by visualising the structure in PyMOL.</p>
</section>
<section id="transfer-the-protein-to-a-larger-box-optional" class="level2">
<h2 class="anchored" data-anchor-id="transfer-the-protein-to-a-larger-box-optional">Transfer the protein to a larger box (optional)</h2>
<p>This may not be required, skip this step if you want to continue with the 10 nm box. We are going to pull the two molecules apart. At the maximum extent, the distance between the two pull groups must be less than half the box size, so a larger box might be necessary. We already have a protein structure that is equilibrated with the force field from the 50 ns unrestrained MD simulation from above. The following method will extract that protein structure and resolvate in a larger box.</p>
<pre><code>cp topol.top topol_50ns.top.bak # create a backup of the topology file.

# extract the protein component, then read it back in again to move it into a larger box 

# using the last frame .gro as both -f and -s. Choose the options to centre on the protein and output the protein only.
gmx trjconv -f md_50ns.gro -s md_50ns.gro -o md_50ns_protein_only.gro -center

# This will set up a larger rectangular box and adjust the orientation.
gmx editconf -f md_50ns_protein_only.gro -o rectbox.gro -center 4 4 2 -rotate 45 45 0  -box 10.0 10.0 13.0</code></pre>
<p>Remove the solvent and ions from the old topology file topol.top by deleting the last 3 lines</p>
<pre><code>SOL         21124
NA               61
CL               60</code></pre>
<p>Then resolvate in 0.15 M NaCl and neutralise</p>
<pre><code>gmx solvate -cp rectbox.gro -cs spc216.gro -o solv2.gro -p topol.top
gmx grompp -f inputs/ions.mdp -c solv2.gro -p topol.top -o ions2.tpr
gmx genion -s ions2.tpr -o solv_ions2.gro -p topol.top -pname NA -nname CL -neutral -conc 0.15

#minimise (less than a minute)
gmx grompp -f inputs/minim.mdp -c solv_ions2.gro -p topol.top -o em2.tpr
gmx mdrun -deffnm em2 -v

#nvt equilibration (less than a minute)
gmx grompp -f inputs/nvt.mdp -c em2.gro -r em2.gro -p topol.top -o nvt2.tpr
gmx mdrun  -deffnm nvt2 -v

#npt equilibration (~2 minutes)
gmx grompp -f inputs/npt.mdp -c nvt2.gro -t nvt2.cpt -r nvt2.gro -p topol.top -o npt2.tpr
gmx mdrun -deffnm npt2 -v</code></pre>
</section>
<section id="pulling-the-tandem-beta-zipper-apart" class="level2">
<h2 class="anchored" data-anchor-id="pulling-the-tandem-beta-zipper-apart">Pulling the tandem beta zipper apart</h2>
<section id="define-the-pull-groups" class="level3">
<h3 class="anchored" data-anchor-id="define-the-pull-groups">Define the pull groups</h3>
<p>We will pull the two molecules apart using a pair of atoms that are close together (0.51 nm), therefore maximizing the pull distance.</p>
<p>Ala44_CA atom 658 on Fn Thr9_CA atom 1438 on the peptide</p>
<section id="section" class="level61">
<p class="heading"></p>
<pre><code>gmx make_ndx -f npt2.gro -o index.ndx 
a 658 &amp; a CA
name 17 Ala44_CA
a 1438 &amp; a CA
name 18 Thr9_CA
q</code></pre>
<p>This creates index.ndx</p>
<p>the pull vector will be 0 0 1</p>
</section>
</section>
<section id="check-the-distances-in-the-npt-equilibration-this-is-a-sanity-check-to-ensure-we-picked-the-correct-atoms" class="level3">
<h3 class="anchored" data-anchor-id="check-the-distances-in-the-npt-equilibration-this-is-a-sanity-check-to-ensure-we-picked-the-correct-atoms">check the distances in the npt equilibration (this is a sanity check to ensure we picked the correct atoms)</h3>
<p>gmx distance -s npt2.tpr -f npt2.trr -n index.ndx -select ‘com of group 17 plus com of group 18’ -oall -oh disthist.xvg -oxyz distxyz.xvg</p>
<p>The average distance: 0.49983 nm</p>
</section>
<section id="add-restraints-to-maintain-the-structure-of-fn" class="level3">
<h3 class="anchored" data-anchor-id="add-restraints-to-maintain-the-structure-of-fn">Add restraints to maintain the structure of Fn</h3>
<p>It is necessary to restrain the Fn protein to prevent rotation. In addition, the force generated from the pull may disrupt the structure of Fn.</p>
<p>Run the command below to see the itp files</p>
<pre><code>ls -ltr *.itp</code></pre>
<p>We need to restrain Chain A (fibronectin), defined by “topol_Protein_chain_A.itp”. Scroll to the end of this file and look for these lines:</p>
<pre><code>; Include Position restraint file
#ifdef POSRES
#include "posre_Protein_chain_A.itp"
#endif</code></pre>
<p>Change the lines to this below:</p>
<pre><code>; Include Position restraint file
#ifdef POSRES_Fn
#include "posre_Protein_chain_A.itp"
#endif</code></pre>
</section>
<section id="edit-the-pull.mdp" class="level3">
<h3 class="anchored" data-anchor-id="edit-the-pull.mdp">Edit the pull.mdp</h3>
<p>Add the -DPOSRES_Fn to the pull.mdp file. It is also necessary to adjust the scaling (as adding restraints with a variable box size may introduce artefacts).</p>
<pre><code>gedit ./inputs/pull.mdp
define    =  -DPOSRES_Fn ;include position restraint on fibronectin
refcoord-scaling = all ; scale each atom's reference coordinates with the pressure-scaling matrix</code></pre>
<p>We need to make some further changes to the pull.mdp file, specifying the two atoms that will be pulled () and the vector to define the direction. The two atoms are Thr9_CA on the peptide Ala44_CA on Fn</p>
<pre><code>; Pull options 
pull                    = yes
pull-ngroups            = 2          ; total number of groups 
pull-ncoords            = 1          ; 
pull-group1-name        = Ala44_CA   ; Fn
pull-group2-name        = Thr9_CA    ; bacterial peptide
pull-coord1-type        = umbrella
pull-coord1-geometry    = direction
pull-coord1-vec         =  0 0 1 ; pull up the cell
pull-coord1-dim         = Y Y Y
pull-coord1-groups      = 1 2
pull-coord1-start       = yes
pull-coord1-rate        = 0.005
pull-coord1-k           = 2000</code></pre>
<p>#pull distance# The pull.mdp has a pull rate of 0.005 nm/ps, which currently over 500ps will extend the vector by 2.5 nm. We need to increase that.</p>
<p>At the maximum extent, the distance between the two pull groups must be less than half the box size (6.5 nm). If you exceed half the box length, it will result in a fatal error. If this happens reduce the pull distance accordingly. We need a margin of error to prevent interactions with neighboring molecules in adjacent cells.</p>
<p>We will use a 5 nm pull.</p>
<p>starting distance + pull distance = maximum extent 0.5 + 5 nm = 5.5 nm</p>
<p>As the pull rate is 0.005 nm/ps, we need to change the time of the simulation to 5/0.005 = 1000 ps. We do that by increasing the number of steps.</p>
<p>1000 ps is 1,000,000 fs</p>
<p>Each step is 2 fs, so 1,000,000/2 = 500000 steps</p>
<p>change this line in pull.mdp</p>
<pre><code>nsteps                  = 500000    ; 1000 ps</code></pre>
</section>
<section id="generate-the-configurations-pulling" class="level3">
<h3 class="anchored" data-anchor-id="generate-the-configurations-pulling">Generate the configurations (pulling)</h3>
<p>This take ~36 minutes</p>
<pre><code>gmx grompp -f inputs/pull.mdp -c npt2.gro -r npt2.gro -t npt2.cpt -n index.ndx -p topol.top -o pull.tpr
gmx mdrun -v -nb gpu -deffnm pull</code></pre>
<p>When you run the first command above, you will see these pull groups. Check the distances are as expected.</p>
<p>Pull group natoms pbc atom distance at start reference at t=0 1 1 0 2 1 0 0.384 nm 0.384 nm</p>
<p>visualise in PyMOL</p>
</section>
</section>
<section id="measure-the-distances-between-the-two-cα-atoms-over-time" class="level2">
<h2 class="anchored" data-anchor-id="measure-the-distances-between-the-two-cα-atoms-over-time">Measure the distances between the two Cα atoms over time</h2>
<p>The command below also generates dist.xvg</p>
<pre><code>gmx distance -s pull.tpr -f pull.xtc -n index.ndx -select 'com of group 17 plus com of group 18' -oall -oh disthist.xvg -oxyz distxyz.xvg

#graph the distance over time
xmgrace dist.xvg
xmgrace -nxy dist.xvg -hdevice PNG -hardcopy -printfile pull_distance.png -noask

#graph the force over time
xmgrace pull_pullf.xvg
xmgrace -nxy pull_pullf.xvg -hdevice PNG -hardcopy -printfile pull_force.png -noask</code></pre>
</section>
<section id="get-the-individual-gro-files-for-umbrella-sampling" class="level2">
<h2 class="anchored" data-anchor-id="get-the-individual-gro-files-for-umbrella-sampling">Get the individual gro files for umbrella sampling</h2>
<p>First, edit the find_windows.py script to use a starting distance of 0.384 nm, ending distance of 5 nm in 0.2 nm increments. (to capture the distance of 5 nm, you may need to enter 5.2 as the maximum)</p>
<pre><code>for d in numpy.arange(0.384, 5.2, 0.2):</code></pre>
<p>Then run the script with this command</p>
<pre><code>python3 find_windows.py</code></pre>
<p>This python script reads the dist.xvg file from the previous step and extracts the .gro file that is closest to each target value. For example:</p>
<pre><code>target value: 3.2 best value: 3.181 at time: 468.0</code></pre>
</section>
<section id="visual-check-of-the-starting-positions-in-pymol" class="level2">
<h2 class="anchored" data-anchor-id="visual-check-of-the-starting-positions-in-pymol">visual check of the starting positions in pymol</h2>
<p>Open these gro files in PyMOL and check that they are as expected. If you are missing the last one at the greatest pull distance then increase the value above and repeat it. Check the extent of the separation. Check at the maximum pull distance there are no interactions with neighboring molecules.</p>
</section>
<section id="run-the-umbrella-sampling" class="level2">
<h2 class="anchored" data-anchor-id="run-the-umbrella-sampling">Run the umbrella sampling</h2>
<p>Edit the template to define the pull groups.</p>
<pre><code>gedit ./inputs/us.tmpl</code></pre>
<p>Edit the lines below, then save this as us.mdp</p>
<p>The options below are for the rectangular cell where we pulled along the Z-axis.</p>
<pre><code>; Pull options
pull                    = yes
pull-ngroups            = 2         ; two groups form one reaction coordinate
pull-ncoords            = 1         ; one reaction coordinate/bias being applied
pull-group1-name        = Ala44_CA   ; Fn
pull-group2-name        = Thr9_CA    ; bacterial peptide
pull-coord1-type        = umbrella  ; harmonic potential
pull-coord1-geometry    = distance
pull-coord1-dim         = N N Y     ; 
pull-coord1-groups      = 1 2
pull-coord1-start       = yes
;pull-coord1-init        = XX 
pull-coord1-rate        = 0.0
pull-coord1-k           = 1000</code></pre>
<p>Edit the ‘run_umbrella_sampling.sh’ script to stop it generating a new .mdp file for each distance. We will use the ‘us.mdp’ for all distances. (comment out lines 17-19, then specify us.mdp on line 22).</p>
<p>Edit the line at the top to define the number of runs for (( i=0; i&lt;=24; i++ ))</p>
<p>The run_umbrella_sampling.sh script will now run generate a new folder called ‘umbrella’, then run npt/nvt equilibration with position restraints on the protein, then a 10 ns unrestrained md simulation in each window.</p>
<pre><code>bash run_umbrella_sampling.sh</code></pre>
<p>#Write the dat files, then run WHAM First, edit line 7 to specify the number of files, in this case 25.</p>
<pre><code>for i in range(0,25):</code></pre>
<p>Then run the script to generate the two .dat files.</p>
<pre><code>python3 write_wham_files.py</code></pre>
<p>Then run wham</p>
<pre><code>gmx wham -it tpr-files.dat -if pullf-files.dat -bsprof -bsres -nBootstrap 200</code></pre>
<p>The output</p>
<pre><code>Warning, poor sampling bin 197 (z=3.36765). Check your histograms!
Warning, poor sampling bin 198 (z=3.38286). Check your histograms!
Warning, poor sampling bin 199 (z=3.39806). Check your histograms!
Initialized rapid wham stuff (contrib tolerance 3.57143e-08)
Evaluating only 1541 of 5600 expressions.
</code></pre>
<p>25 Kj/mol</p>
<p>xmgrace -nxy histo.xvg xmgrace profile.xvg xmgrace bsResult.xvg</p>
<p>To calculate the Gibbs free energy change (ΔG) from a dissociation constant (Kd), use: <span class="math display">\[
\Delta G = RT \ln(K_d)
\]</span> Where: R is the gas constant (8.314 J⋅K⁻¹⋅mol⁻¹) T is the temperature (298 K) <code>$K_d$</code> is the dissociation constant in moles per litre (<span class="math inline">\(3.6\,\mu\mathrm{M}\)</span> is 5×10−7 M)</p>
<p>Lets plot that bootstrap graph.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="co">#expected from ITC</span></span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>(<span class="fl">8.314</span><span class="sc">*</span><span class="dv">298</span><span class="sc">*</span><span class="fu">log</span>(<span class="fl">3.6</span><span class="sc">*</span><span class="dv">10</span><span class="sc">^-</span><span class="dv">6</span>) )<span class="sc">/</span><span class="dv">1000</span> <span class="co"># in Kj</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] -31.05532</code></pre>
</div>
</div>
<p>the profile suggests 85 Kj/mol</p>
<pre><code>library(tidyverse)
library(here)

#I tried the read_xvg, but it did not work, so we manually read the file in and extract the numbers

raw &lt;- readLines(here("gmx_tutorials_jpcb","inputs","04_FnBPA1", "bsResult.xvg"))

# keep only numeric lines
dat &lt;- raw[grepl("^[[:space:]]*[0-9]", raw)]

bs_profile &lt;- read.table(text = dat)
colnames(bs_profile) &lt;- c("x_nm", "meanEnergy", "stddev")
global_minimum &lt;- min(bs_profile$meanEnergy) #find the global minimum

bs_profile &lt;- bs_profile |&gt;  mutate(meanEnergy = meanEnergy-global_minimum)

bs_profile |&gt; ggplot()+
  aes(x= x_nm, y=meanEnergy)+
  geom_point()+
  ylab("Free Energy (KJ/mol)")+
  xlab("Reaction Coordinate (nm)")+
  geom_text(aes(x = 1, y = 33, label = "31.9 KJ/mol"), 
  colour = "blue",
  size = 4)+
  geom_hline(yintercept =  31.94637, color = "blue")
</code></pre>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
        const codeEl = trigger.previousElementSibling.cloneNode(true);
        for (const childEl of codeEl.children) {
          if (isCodeAnnotation(childEl)) {
            childEl.remove();
          }
        }
        return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp('/' + window.location.host + '/');
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->




</body></html>